### PR TITLE
Initial support for Document, EventTarget, NodeList and Iterator 

### PR DESCRIPTION
**Pull request by jonathanKingston**
@alexcrichton https://github.com/rustwasm/wasm-bindgen/compare/master...jonathanKingston:web-sys/document-support?expand=1#diff-9f9955b41f2425125eaf42c53a05c94dR558 is the mentioned issue from https://github.com/rustwasm/wasm-bindgen/issues/514#issuecomment-407090466

The other question I had was surrounding: `element.query_selector_all('x').get(0)` returning a `Node` not an `Element`.  Is this something https://github.com/rustwasm/rfcs/pull/2 will tackle?

### CODE REVIEW

**Comment by alexcrichton** [comment]
Awesome, thanks for this!

I've switched the `throws` there to `false` and I think I see the issue! The first problem is that `Option` wasn't defined in [this list](https://github.com/rustwasm/wasm-bindgen/blob/cefe1681ab37091ca07080fe1234390e5b5aff47/crates/webidl/src/lib.rs#L83-L86). After that though it looked like the path traversal code wasn't handling generics so there were lots of compilations errors. After [filling out the missing cases](https://gist.github.com/alexcrichton/7a74a6781ad7bea5e9974f3ec3aed1f3) though it should be all good to go!

File: crates/webidl/src/lib.rs [hunk 1]
```diff
@@ -520,6 +522,48 @@ impl<'a> WebidlParse<&'a str> for webidl::ast::RegularAttribute {
...
+            return Ok(());
+        }
+
+        let throws = util::throws(&self.extended_attributes);
+        let return_value = webidl::ast::ReturnType::NonVoid(self.value_type.clone());
+        let args = [];
+        first_pass
+            .create_basic_method(
+                &args,
+                Some(&"values".to_string()),
```
Comment (alexcrichton) [inline crates/webidl/src/lib.rs]: I wasn't quite sure what these two methods were doing, is this what it means to be iterable? (a `values` and an `item` method?)

Comment (jonathanKingston) [reply crates/webidl/src/lib.rs]: It comes from the `iterable` in webidl yeah: https://heycam.github.io/webidl/#es-iterable

I couldn't get `forEach` to work as we don't have callbacks yet.

`entries`, `keys` and `values` and some other behaviour is needed.

I messed up though `item` is a getter on the webidl which isn't supported yet.

**Comment by jonathanKingston** [comment]
I filled in the `Parenthesized` `PathArguments` and it looks like it is working (I have more in Element now like `getAttribute` which wasn't working before).

I commented out the code for now as I think I need to implement a trait for iterator similar to how `OptionFromWasmAbi` and `OptionIntoWasmAbi` work.

I think perhaps I could remove the function comments and it should be merged as is (without iterator).

File: crates/webidl/src/lib.rs (hunk 1 above)
Comment (jonathanKingston) [reply crates/webidl/src/lib.rs]: Thanks for this direction, it really helped me understand the codebase a lot more too!

File: crates/backend/src/defined.rs [hunk 2]
```diff
@@ -147,19 +161,74 @@ impl ImportedTypes for syn::TypePath {
...
+                for arg in data.args.iter() {
+                    arg.imported_types(f);
+                }
+            }
+//TOCHECK
+            syn::PathArguments::Parenthesized(data) => {
+                for input in data.inputs.iter() {
+                    input.imported_types(f);
+                }
+                // TODO do we need to handle output here?
```
Comment (alexcrichton) [inline crates/backend/src/defined.rs]: I think so yeah if you get a chance, makes sense to go ahead and add it here!

**Comment by alexcrichton** [comment]
Awesome looks great to me!

Do you think we'll need a new trait for the iterator types? I'm actually curious if we can get by with an `Iterator` trait implementation? Or perhaps things like `values` return `impl Iterator`?

**Comment by jonathanKingston** [comment]
Wow cool thanks for merging!

> Do you think we'll need a new trait for the iterator types? I'm actually curious if we can get by with an Iterator trait implementation? Or perhaps things like values return impl Iterator?

More experimenting needed I think, I am hoping we can use that but I'm still learning here.

**Comment by alexcrichton** [comment]
Sure thing! I'm all for incremental changes anyway :)

**Pull request merged**